x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/aggs/correlation/BucketCorrelationAggregationBuilder.java

```diff
@@ -0,0 +1,130 @@
+            FUNCTION
+        );
+    }
+
+    private final CorrelationFunction correlationFunction;
+
+    public BucketCorrelationAggregationBuilder(
+        String name,
+        String bucketsPath,
+        CorrelationFunction correlationFunction
+    ) {
+        super(
+            name,
+            NAME.getPreferredName(),
+            new String[] {bucketsPath}
+        );
+        this.correlationFunction = correlationFunction;
+    }
+
+    public BucketCorrelationAggregationBuilder(StreamInput in) throws IOException {
+        super(in, NAME.getPreferredName());
+        this.correlationFunction = in.readNamedWriteable(CorrelationFunction.class);
+    }
+
+    @Override
+    public String getWriteableName() {
+        return NAME.getPreferredName();
+    }
+
+    @Override
+    protected void doWriteTo(StreamOutput out) throws IOException {
+        out.writeNamedWriteable(correlationFunction);
+    }
+
+    @Override
+    protected PipelineAggregator createInternal(Map<String, Object> metadata) {
+        return new BucketCorrelationAggregator(name, correlationFunction, bucketsPaths[0], metadata);
+    }
+
+    @Override
+    protected boolean overrideBucketsPath() {
+        return true;
+    }
+
+    @Override
+    protected XContentBuilder internalXContent(XContentBuilder builder, Params params) throws IOException {
+        builder.field(BUCKETS_PATH_FIELD.getPreferredName(), bucketsPaths[0]);
+        NamedXContentObjectHelper.writeNamedObject(builder, params, FUNCTION.getPreferredName(), correlationFunction);
+        return builder;
+    }
+
+    @Override
+    protected void validate(ValidationContext context) {
+
+        final String firstAgg = bucketsPaths[0].split("[>\\.]")[0];
+        Optional<AggregationBuilder> aggBuilder = context.getSiblingAggregations().stream()
+            .filter(builder -> builder.getName().equals(firstAgg))
+            .findAny();
+        if (aggBuilder.isEmpty()) {
+            context.addBucketPathValidationError("aggregation does not exist for aggregation [" + name + "]: " + bucketsPaths[0]);
+            return;
+        }
+        AggregationBuilder aggregationBuilder = aggBuilder.get();
+        if (aggregationBuilder.bucketCardinality() != AggregationBuilder.BucketCardinality.MANY) {
+            context.addValidationError("The first aggregation in " + PipelineAggregator.Parser.BUCKETS_PATH.getPreferredName()
+                + " must be a multi-bucket aggregation for aggregation [" + name + "] found :"
+                + aggBuilder.get().getClass().getName() + " for buckets path: " + bucketsPaths[0]);
+            return;
+        }
+        correlationFunction.validate(context, bucketsPaths[0]);
+    }
+
+    @Override
+    public boolean equals(Object o) {
+        if (this == o) return true;
+        if (o == null || getClass() != o.getClass()) return false;
+        if (super.equals(o) == false) return false;
```

Comment:
`super.equals` already does the reference and class comparison so you can probably skip that bit.